### PR TITLE
TermCont improvements

### PR DESCRIPTION
**Pull request by srid**
Change the TermCont type to be parameterized by the result type `r`, viz.: `TermCont r s a`.  This enables writing utility functions for TermCont monad in smart contracts, for example:

```haskell
-- | Terminate with given value on empty list, otherwise continue.
whenNonEmpty :: Term s r -> PList a s -> TermCont r s (Term s a, Term s (PList a))
whenNonEmpty x0 list = TermCont $ \k ->
  case list of
    PSCons x xs -> k (x, xs)
    PSNil -> x0
```

Other changes:
- Introduce `runTermContId`, because using `flip runTermCont id` _inline_ fails type-check.
- Add `tcont` alias to `TermCont`, analogous to `cont` of `ContT` (see examples/../Api.hs)
- Add examples, from #224 
- #225 
    - Add `MonadFail` instance
    - Re-organize module, to avoid cyclic imports.
    - Add benchmarks for all 4 types of syntax.

### CODE REVIEW

**Comment by srid** [comment]
I assume backwards compatibility is not important here, because this goes to `staging`.

**Comment by L-as** [comment]
I can't know that, but it's pretty likely.

**Comment by srid** [comment]
@TotallyNotChase We are using TermCont in our smart contract. The main migration change required, when using this PR, is to add the extra `r` type argument; ie., rename all occurrences of `TermCont s` to `TermCont r s`.

**Comment by L-as** [comment]
I'm fixing the backward compatibility in this PR.

**Comment by L-as** [comment]
@srid I fixed the issues I had. @TotallyNotChase Can you review?

**Comment by TotallyNotChase** [comment]
Does this work fine on GHC 8? Including all the utility functions? I'll check the compilation, at the very least, quickly.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,13 @@
 # Revision history for plutarch
 
+# Unreleased
+
+- `TermCont`: Parametrize by result type; add `MonadFail` instance; etc.
+
+  Also, export from `Plutarch.TermCont`, and then from `Plutarch.Prelude` (TermCont is no longer exported by `Plutarch.Internal`).
```
Comment (TotallyNotChase) [inline CHANGELOG.md:7]: It's also exported from `Plutarch` now. I don't think the "(TermCont is no longer exported by `Plutarch.Internal`)" part is really needed. Users shouldn't care about restructures of internal modules.

File: examples/Examples/Api.hs [hunk 2]
```diff
@@ -171,19 +171,36 @@ checkSignatory = plam $ \ph ctx' ->
...
+              -- Signature not present.
+              perror
+      _ ->
+        ptraceError "checkSignatoryCont: not a spending tx"
+
+-- | `checkSignatory` implemented using `runTermCont`
+checkSignatoryTermCont :: Term s (PPubKeyHash :--> PScriptContext :--> PUnit)
+checkSignatoryTermCont = plam $ \ph ctx' -> unTermCont $ do
+  ctx <- tcont $ pletFields @["txInfo", "purpose"] ctx'
+  PSpending _ <- tcont (pmatch . pfromData $ ctx.purpose)
```
[the remark this answers is not in the document]
Comment (L-as) [reply examples/Examples/Api.hs:195]: I don't think it's worth it.

**Comment by L-as** [comment]
CI checks GHC 8.10.7 too.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,13 @@
 # Revision history for plutarch
 
+# Unreleased
+
+- `TermCont`: Parametrize by result type; add `MonadFail` instance; etc.
```
Comment (TotallyNotChase) [inline CHANGELOG.md:5]: Can we have a bit more info on here? It's not actually explicitly parameterized by result type now.

**Review by TotallyNotChase — CHANGES_REQUESTED** [review]
Just the changelog needs change hehe

The example with `TermCont` does indeed work in GHC 8, rejoice! I'll add this to the guide as well of course.

Can we retarget this to `master` please?

**Comment by srid** [comment]
Let me review this too, please don't merge yet.

**Comment by srid** [comment]
@L-as Your change to `TermCont` put us back to square 1, and voids the "This enables writing utility functions for TermCont monad in smart contracts" in the PR description. 

```
examples/Examples/Api.hs:211:13-14: error:
    • Couldn't match type ‘r1’ with ‘r’
      Expected: Term s r
        Actual: Term s r1
```

Specifically, how would you make this type-check? 

```diff
diff --git a/examples/Examples/Api.hs b/examples/Examples/Api.hs
index 9c14f48..17dacd0 100644
--- a/examples/Examples/Api.hs
+++ b/examples/Examples/Api.hs
@@ -194,6 +194,7 @@ checkSignatoryTermCont = plam $ \ph ctx' -> unTermCont $ do
   ctx <- tcont $ pletFields @["txInfo", "purpose"] ctx'
   PSpending _ <- tcont (pmatch . pfromData $ ctx.purpose)
   let signatories = pfield @"signatories" # ctx.txInfo
+  (_, _) <- whenNonEmpty perror =<< tcont (pmatch $ pfromData signatories)
   pure $
     pif
       (pelem # pdata ph # pfromData signatories)
@@ -202,6 +203,13 @@ checkSignatoryTermCont = plam $ \ph ctx' -> unTermCont $ do
       -- Signature not present.
       perror
 
+-- | Terminate with given value on empty list, otherwise continue.
+whenNonEmpty :: Term s r -> PBuiltinList a s -> TermCont s (Term s a, Term s (PBuiltinList a))
+whenNonEmpty x0 list = TermCont $ \k ->
+  case list of
+    PCons x xs -> k (x, xs)
+    PNil -> x0
+
 getFields :: Term s (PData :--> PBuiltinList PData)
 getFields = phoistAcyclic $ plam $ \addr -> psndBuiltin #$ pasConstr # addr
```

**Comment by L-as** [comment]
@srid You haven't specified the `r`? `whenNonEmpty :: Term s r -> PBuiltinList a s -> TermCont @r s (Term s a, Term s (PBuiltinList a))`

File: CHANGELOG.md (hunk 3 above)
Comment (L-as) [reply CHANGELOG.md:5]: I think this wording is fine.

File: CHANGELOG.md (hunk 1 above)
Comment (L-as) [reply CHANGELOG.md:7]: The internal stuff isn't necessary yeah. I don't think the other stuff needs to change though.

File: Plutarch/TermCont.hs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  tcont,
+) where
+
+import Data.Kind (Type)
+import Data.String (fromString)
+import Plutarch.Internal (Dig, PType, S, Term (Term), asRawTerm, getTerm, hashRawTerm)
+import Plutarch.Trace (ptraceError)
+
+newtype TermCont :: forall (r :: PType). S -> Type -> Type where
+  TermCont :: forall r s a. {runTermCont :: ((a -> Term s r) -> Term s r)} -> TermCont @r s a
```
Comment (srid) [inline Plutarch/TermCont.hs:17]: TIL!

What is this syntax? I had no idea GHC supported implicit type parameter on (new)types, that you can pass explicitly using `TypeApplications`.

Comment (L-as) [reply Plutarch/TermCont.hs:17]: It's just GADT syntax with `newtype`.

Comment (TotallyNotChase) [reply Plutarch/TermCont.hs:17]: I'm also learning so much just by observing Plutarch :D

Comment (srid) [reply Plutarch/TermCont.hs:17]: For clarity, and foresaking newtype, this is equivalent to:

```haskell
type TermCont :: forall (r :: PType). S -> Type -> Type
data TermCont s a where
  TermCont :: forall r s a. {runTermCont :: ((a -> Term s r) -> Term s r)} -> TermCont @r s a
```

It is the `TermCont @r` part that I had no idea could be done in GHC; ie. implicit type parameters. The GADT unified the two `r`'s, which is a neat trick. So we are using GADT to workaround lack of dependent types.

**Comment by srid** [comment]
> @srid You haven't specified the `r`? `whenNonEmpty :: Term s r -> PBuiltinList a s -> TermCont @r s (Term s a, Term s (PBuiltinList a))`

Yea, I had overlooked the implicit parameter. It works now! I approve the PR, @L-as

**Review by TotallyNotChase — APPROVED** [review]
GHC 8 users rejoice!

File: Plutarch/TermCont.hs (hunk 4 above)
Comment (L-as) [reply Plutarch/TermCont.hs:17]: Essentially yeah this is the idea.

**Pull request merged**